src/HIE/Bios/Types.hs

```diff
@@ -44,21 +44,60 @@ data ActionName a
   | Other a
   deriving (Show, Eq, Ord, Functor)
 
-data CradleAction a = CradleAction {
-                        actionName    :: ActionName a
-                      -- ^ Name of the action.
-                      , runCradle     :: LoggingFunction -> FilePath -> IO (CradleLoadResult ComponentOptions)
-                      -- ^ Options to compile the given file with.
-                      , runGhcCmd :: [String] -> IO (CradleLoadResult String)
-                      -- ^ Executes the @ghc@ binary that is usually used to
-                      -- build the cradle. E.g. for a cabal cradle this should be
-                      -- equivalent to @cabal exec ghc -- args@
-                      }
+data CradleAction a = CradleAction
+  { actionName    :: ActionName a
+  -- ^ Name of the action.
+  , runCradle     :: LoggingFunction -> FilePath -> IO (CradleLoadResult LoadResult)
+  -- ^ Options to compile the given file with.
+  --
+  -- The given FilePath /must/ be part of 'LoadResult.loadResultComponent' if
+  -- the loading operation succeeds.
+  , runGhcCmd :: [String] -> IO (CradleLoadResult String)
+  -- ^ Executes the @ghc@ binary that is usually used to
+  -- build the cradle. E.g. for a cabal cradle this should be
+  -- equivalent to @cabal exec ghc -- args@
+  }
   deriving (Functor)
 
 instance Show a => Show (CradleAction a) where
   show CradleAction { actionName = name } = "CradleAction: " ++ show name
 
+type LoadResult = LoadResult' ComponentOptions
+
+-- | Record for expressing successful loading.
+-- Can express partial success.
+data LoadResult' a = LoadResult
```

Comment:
What about somethin like

```haskell
data ComponentInfo = ComponentInfo
  { componentOptions :: Maybe a
   componentDependencies: [a]
  }
```

It would be generic enough to allow add more component information afterwards

Reply:
Hm, maybe we can drop the prefix somehow to be less "Component"-y? However, since the record field prefix will probably still be required, probably pointless.

Reply:
yeah, `Info` sounds good too